mip/driver_imx_rt1020.c

```diff
@@ -0,0 +1,302 @@
+}
+
+static uint32_t eth_read_phy(uint8_t addr, uint8_t reg) {
+  ENET->EIR |= BIT_SET(23); // MII interrupt clear
+  uint32_t mask_phy_adr_reg = 0x1f; // 0b00011111: Ensure we write 5 bits (Phy address & register)
+  uint32_t phy_transaction = 0x00;
+  phy_transaction = (0x1 << 30) \
+                  | (0x2 << 28) \
+                  | ((uint32_t)(addr & mask_phy_adr_reg) << 23) \
+                  | ((uint32_t)(reg & mask_phy_adr_reg)  << 18) \
+                  | (0x2 << 16);
+
+  //MG_INFO(("phy_transaction %x", phy_transaction));
+  ENET->MMFR = phy_transaction;
+  wait_phy_complete();
+
+  // volatile uint32_t phy_read_out = ENET->MMFR;
+  //MG_INFO(("phy_read_out 0x%x", phy_read_out));
+  return (ENET->MMFR & 0x0000ffff);
+}
+
+static void eth_write_phy(uint8_t addr, uint8_t reg, uint32_t val) {
+  ENET->EIR |= BIT_SET(23); // MII interrupt clear
+  uint8_t mask_phy_adr_reg = 0x1f; // 0b00011111: Ensure we write 5 bits (Phy address & register)
+  uint32_t mask_phy_data = 0x0000ffff; // Ensure we write 16 bits (data)
+  addr &= mask_phy_adr_reg;
+  reg &= mask_phy_adr_reg;
+  val &= mask_phy_data;
+  uint32_t phy_transaction = 0x00;
+  phy_transaction = (uint32_t)(0x1 << 30) \
+                  | (uint32_t)(0x1 << 28) \
+                  | (uint32_t)(addr << 23) \
+                  | (uint32_t)(reg  << 18) \
+                  | (uint32_t)(0x2 << 16) \
+                  | (uint32_t)(val);
+  ENET->MMFR = phy_transaction;
+  wait_phy_complete();
+}
+
+// FEC RX/TX descriptors (Enhanced descriptor not enabled)
+// Descriptor buffer structure, little endian
+
+typedef struct enet_bd_struct_def
+{
+    uint16_t length;  // Data length
+    uint16_t control; // Control and status
+    uint32_t *buffer; // Data ptr
+} enet_bd_struct_t;
+
+// Descriptor and buffer globals, in non-cached area, 64 bits aligned.
+
+__attribute__((section("NonCacheable,\"aw\",%nobits @"))) enet_bd_struct_t rx_buffer_descriptor[(ENET_RXBD_NUM)] __attribute__((aligned((64U))));
+__attribute__((section("NonCacheable,\"aw\",%nobits @"))) enet_bd_struct_t tx_buffer_descriptor[(ENET_TXBD_NUM)] __attribute__((aligned((64U))));
+
+uint8_t rx_data_buffer[(ENET_RXBD_NUM)][((unsigned int)(((ENET_RXBUFF_SIZE)) + (((64U))-1U)) & (unsigned int)(~(unsigned int)(((64U))-1U)))] __attribute__((aligned((64U))));
+uint8_t tx_data_buffer[(ENET_TXBD_NUM)][((unsigned int)(((ENET_TXBUFF_SIZE)) + (((64U))-1U)) & (unsigned int)(~(unsigned int)(((64U))-1U)))] __attribute__((aligned((64U))));
+
+// Initialise driver imx_rt1020
+
+static bool mip_driver_imx_rt1020_init(uint8_t *mac, void *data) {
+  // ENET Reset, wait complete
+  ENET->ECR |= BIT_SET(0);
+  while((ENET->ECR & BIT_SET(0)) != 0) {}
+
+  // Re-latches the pin strapping pin values
+  ENET->ECR |= BIT_SET(0);
+  while((ENET->ECR & BIT_SET(0)) != 0) {}
+
+  // Setup MII/RMII MDC clock divider (<= 2.5MHz).
+  ENET->MSCR = 0x130; // HOLDTIME 2 clk, Preamble enable, MDC MII_Speed Div 0x30
```

Comment:
This is clock speed dependent

Reply:
Yes so far the clock is only 50 MHz, there's peripheral clock being used in there.
I'll have to work on this to be more general. Though it should be working all right for the 10/100M setup.